src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
+): Promise<T> {
+  if (nodeCache.has(entity)) {
+    return nodeCache.get(entity);
+  }
+
+  const cache = new Map();
```

Comment:
What's the intended difference between `nodeCache` and `cache`? I can't see `cache` being used anywhere?

Reply:
nodeCache is a map of visited entities, the other cache is the map of cached conditionals, I guess this can be combined? I mainly included it because getAndCacheDependency needed it